go-controller/pkg/ovn/egressgw.go

```diff
@@ -143,23 +141,33 @@ func (oc *Controller) deleteRouteInfoLocked(name ktypes.NamespacedName) *externa
 	// we try to lock the routeInfo to avoid any deadlock
 	oc.exGWCacheMutex.RLock()
 	routeInfo := oc.externalGWCache[name]
+	klog.Warningf("##### [%p %s] deleteRouteInfoLocked() got routeInfo", routeInfo, name)
 	oc.exGWCacheMutex.RUnlock()
 
 	if routeInfo == nil {
 		return nil
 	}
 	routeInfo.Lock()
 
-	// ensure the routeInfo we acquired the lock for is consistent with the
-	// one still in the cache
-	oc.exGWCacheMutex.Lock()
-	defer oc.exGWCacheMutex.Unlock()
-	if routeInfo != oc.externalGWCache[name] {
+	if routeInfo.deleted {
+		klog.Warningf("##### [%p %s] deleteRouteInfoLocked() routeInfo already deleted", routeInfo, name)
 		routeInfo.Unlock()
 		return nil
 	}
 
-	delete(oc.externalGWCache, name)
+	klog.Warningf("##### [%p %s] deleteRouteInfoLocked() marking routeInfo deleted", routeInfo, name)
+	routeInfo.deleted = true
+
+	go func() {
+		oc.exGWCacheMutex.Lock()
+		defer oc.exGWCacheMutex.Unlock()
+		klog.Warningf("##### [%p %s] deleteRouteInfoLocked() checking routeInfo for removal", routeInfo, name)
+		if newRouteInfo := oc.externalGWCache[name]; routeInfo == newRouteInfo {
+			klog.Warningf("##### [%p %s] deleteRouteInfoLocked() removing routeInfo", routeInfo, name)
+			delete(oc.externalGWCache, name)
```

Comment:
so this makes sure we first check if something changed between acquiring `routeInfo` lock and releasing `exGWCacheMutex` lock. Why did we convert this to a goroutine? It's not guaranteed to finish before we exit out of `deleteRouteInfoLocked` right? So will we get into trouble if this routine delays finishing and we hit `delete(routeInfo.podExternalRoutes, pod)` in `deleteGWRoutesForPod` which could be nil at that point?

Reply:
For the goroutine question, the problem is that we cannot attempt to acquire the cache mutex while also holding the routeInfo lock, or we risk the deadlock situation. Punting the actual removal to a goroutine breaks that chain and allows us to remove it slightly later, at the cost of possibly having issues with quick create/delete/create situations; but that's easily worked around by deleting the pod, waiting a couple seconds, and restarting. I'd argue that's preferable to a deadlock, until we convert this to a workqueue.

For the section question about delays, routeInfo is still a reference to the object that we're removing from the cache, and the object is still alive since we're passing the reference around. It just doesn't have an entry in the cache any more after the goroutine is done. But the reference gets passed back to `deleteGWRoutesForPod()`and will be alive until that function exits. So the `delete(routeInfo.podExternalRoutes, pod)` will still be working with a valid routeInfo object.